Migrate to pyweatherflowudp for WeatherFlow sensor communication

natekspencer opened the pull request:
This moves the communication from WeatherFlow sensors out of this project and, instead, uses the pyweatherflowudp package that was rewritten.

This allows for multiple sensors to now individually be tracked

Breaking Changes:
- status sensor is now a timestamp (referencing the up_since timestamp of the device) instead of the "humanized" time string since HA takes care of "humanizing" on the front end. This reduces state updates on the sensor since it doesn't have to update every time the uptime seconds change
- `<device>_status` (where device is hub, air, sky or tempest) is now just `status`
- similarly, `battery_<sensor>`, `battery_level_<sensor>` and `voltage_<sensor>` are now just `battery`, `battery_level` and `voltage`, respectively

Other Changes:
- Multiple devices are now created in mqtt (one for each device)
- Removes the `TEMPEST_DEVICE` environment variable/config option since we no longer need a user to tell us the type of device

natekspencer commented:
The breaking/other changes listed above are the user facing changes. In the code, you'll see a few other changes. The biggest one is the migration of the sensor metadata out of the const.py and into sensor_descriptions.py, which utilizes a class structure instead of an ordered list. This should make it easier to manage the list of sensors going forward without having to specify redundant or unneeded values just to ensure the order of the list is maintained.

Let me know if you have questions!

briis commented:
Thanks Nate,
I will go through the code today.

natekspencer commented:
Additionally, this should solve https://github.com/briis/hass-weatherflow2mqtt/issues/66 by creating multiple devices with their own state and attribute topics.

natekspencer commented on weatherflow2mqtt/weatherflow_mqtt.py:
```diff
@@ -550,21 +549,11 @@ def _setup_sensors(self, device: WeatherFlowDevice) -> None:
-                    sensor_event,
-                )
-                continue
-
```
The weather sensor was moved out of DEVICE_SENSORS, so this code would never fire

natekspencer commented on weatherflow2mqtt/weatherflow_mqtt.py:
```diff
@@ -550,21 +549,11 @@ def _setup_sensors(self, device: WeatherFlowDevice) -> None:
             ):
-                _LOGGER.debug(
-                    "Skipping %s sensor since it is missing from %s", sensor_id, device
-                )
```
Produces a lot of noise for non-Tempest devices, so just removed it

natekspencer commented on weatherflow2mqtt/weatherflow_mqtt.py:
```diff
@@ -166,19 +176,346 @@ async def connect(self) -> None:
+        """Construct and return a sensor payload."""
+        payload = OrderedDict()
+
+        payload["name"] = f"WF {sensor.name}"
```
Do we need to prefix sensor names with `WF `? I left it in because my main goal was implementing the pyweatherflowudp package. So most all of the changes were related in getting that to work well while trying to keep a lot of other things the same.

briis replied:
Yes I think we still need that. First of all, it would be a major breaking change if we took it out, and secondly, there are so many Weather integrations out there, that separating where the data comes from, is very useful.

natekspencer replied:
Maybe we could change it to be the device name/serial instead of "WF " to match other HA component standards in the future.

The only reason I suggest this is because the HA frontend will hide the first part of the sensor name that matches the device name. So if a device is named "Hub HB-0001" and the sensor is named "Hub HB-0001 Status", it will display as just "Status" on the Hub's device page, which is nice since it removes the irrelevant/duplicative info to help you see what you really want to.

But for now I'll leave it.

briis replied:
I like your idea. Leave as is for now, and then I will put a *deprecated* message in the release notes. It can then be changed on the next release.

natekspencer commented on weatherflow2mqtt/weatherflow_mqtt.py:
```diff
@@ -166,19 +176,346 @@ async def connect(self) -> None:
+            "identifiers": [f"WeatherFlow_{serial_number}"],
+            "connections": [["mac", serial_number]],
+            "manufacturer": "WeatherFlow",
+            "name": "WeatherFlow2MQTT",
+            "model": f"WeatherFlow Weather Station V{VERSION}",
```
As above, should we put the actual device model here for name and model? It would help to distinguish the devices under the mqtt integration.

briis replied:
Yes, lets do that.

natekspencer commented on weatherflow2mqtt/weatherflow_mqtt.py:
```diff
@@ -166,19 +176,346 @@ async def connect(self) -> None:
+        self._queue = asyncio.Queue()
+        self._queue_task = asyncio.ensure_future(self._mqtt_queue_processor())
+
+        fcst_state_topic = MQTT_TOPIC_FORMAT.format(DOMAIN, FORECAST_ENTITY, "state")
+        fcst_attr_topic = MQTT_TOPIC_FORMAT.format(
+            DOMAIN, FORECAST_ENTITY, "attributes"
+        )
+        for sensor in FORECAST_SENSORS:
+            discovery_topic = MQTT_TOPIC_FORMAT.format(DOMAIN, sensor.id, "config")
+            payload: OrderedDict | None = None
+            if self.forecast is not None:
+                payload = self._get_sensor_payload(
+                    sensor, DOMAIN, fcst_state_topic, fcst_attr_topic, VERSION
+                )
+            self._add_to_queue(
+                discovery_topic, json.dumps(payload or {}), qos=1, retain=True
+            )
```
The way this is now, it will currently end up under its own unique device since a forecast is not really associated with a sensor. We could, instead, attach it to the hub since the hub is roughly equivalent to a station. But if there are multiple hubs on one network (huge edge case, but still possible), which hub would be appropriate? Or just add it to both in that scenario?

^ Those were my thoughts going through it, which is why I ended up creating it on its own device.

briis replied:
I am actually not sure what happens if you have more than 1 Hub. They both broadcast on the same port, so which one would be picked up? I have 2 Hubs available (Only 1 operational right now) so I will test it as soon as I have time to set it up.

But yes, attach the forecast to the Hub, that makes most sense.

briis replied:
Just fired up a container with the changes, and I am not getting any forecast data, so for some reason, which I cannot figure out, it does not seem to fire the event, despite the Vars being set in the Docker command.

briis replied:
CORRECTION: After a few Restarts the Forecast data is now populating. So no error here.

briis replied:
Final Comment here: Could you add a log entry so that people can see that the Forecast sensor is created. I guess that was part of why I missed that it was there. Currently it only shows for the other sensors.

natekspencer replied:
> They both broadcast on the same port, so which one would be picked up?

Both hubs _should_ be picked up in that scenario. While they both broadcast via the same port, they have a unique IP address which allows them to operate independently. There can only be one listener attached to an address/port combo however, which is the "0.0.0.0"/50222 address/port (essentially a non-specific address so it can just listen for these datagrams that are announced on the network via that port).

natekspencer replied:
> Final Comment here: Could you add a log entry so that people can see that the Forecast sensor is created. I guess that was part of why I missed that it was there. Currently it only shows for the other sensors.

Done

natekspencer commented on weatherflow2mqtt/weatherflow_mqtt.py:
```diff
@@ -166,19 +176,346 @@ async def connect(self) -> None:
+
+                    # Check for a custom function
+                    if (fn := sensor.custom_fn) is not None:
+                        # TODO: Handle unique data points more elegantly
+                        if sensor.id == "feelslike":
+                            attr = fn(self.cnv, device, self.wind_speed)
+                        elif sensor.id in ("sealevel_pressure", "visibility"):
+                            attr = fn(self.cnv, device, self.elevation)
+                        elif sensor.id == "wbgt":
+                            attr = fn(self.cnv, device, self.solar_radiation)
```
You'll notice a few "TODO"s because I haven't figured out the best way to handle these yet while I was working on everything else. I wanted to leave a note in it so that when I see it I can try to continue thinking through it.

briis replied:
I have some "real" work for the next hours, so it will take a bit before I get to testing it.

natekspencer replied:
We can address the TODOs in another release

natekspencer commented on DOCS.md:
```diff
@@ -10,17 +10,13 @@ To install the add-on, first follow the installation steps from the [README on G
 
 ## Configuration
 
-### Option: `TEMPEST_DEVICE`: (default: True)
-
-If you have a Tempest Weather Station set this to True. If False, the program will assume you have the older AIR and SKY units. Default is *True*
-
```
The `TEMPEST_DEVICE` variable is no longer needed

natekspencer commented on README.md:
```diff
@@ -14,12 +14,12 @@ There is support for both the AIR & SKY devices and the TEMPEST device.
+   1. [Docker Volume](#docker-volume)
+   2. [Docker Environment Variables](#docker-environment-variables)
 3. [Available Sensors](#available-sensors)
-    1. [Sensor Structure](#sensor-structure)
-    2. [High and Low Values](#high-and-low-values)
-    3. [Long Term Statistics](#Home-Assistant-Long-Term-Statistics)
+   1. [Sensor Structure](#sensor-structure)
+   2. [High and Low Values](#high-and-low-values)
+   3. [Long Term Statistics](#Home-Assistant-Long-Term-Statistics)
```
My markdown formatter is apparently setup quite a bit differently than yours is. We probably need to upload a file that tells various editors how to format this (and other file types) so we have consistency when saving.

briis replied:
I have not done anything specific on that - I use Visual Studio code on a Mac, to do everything. But whatever you can think of to align here, I will implement.

natekspencer replied:
Sweet, that's what I use as well. I'll see what I can add to the settings or files to give everyone the same output when saving in another PR

briis commented:
The container has now been running for about 30 min. Data is coming in as expected and I see no errors in the Log. As said above, the only thing that does not seem to work is the Forecast.

I have 3 devices under the MQTT Integration relating to this:
- 1 for the Hub Status
- 1 for Weather Forecast (No data)
- 1 for the Tempest Sensors

This is what I expected based on your description.

I will now turn on Debug and let it run for a while to see if anything pops up here.

briis commented:
LGTM👍🏻

natekspencer commented:
> * 1 for Weather Forecast (No data)

This should now be associated with the Hub instead of its own device. You'll need to delete that device first before it shows correctly when re-running

briis commented:
I am running the test right now. Weather is now associated with the Hub 👍🏻

I deleted everything from MQTT, and have now a Hub and a Tempest Device. 
They are both placed in an Area called *Backyard*. Where does this come from? I don't have such an area defined. Should be left empty for people to fill themself.

All sensors, except Absolute Humidity, are not associated with the Tempest device. They are in the system, but as separate sensors. This needs to be fixed.

natekspencer commented:
> They are both placed in an Area called _Backyard_. Where does this come from? I don't have such an area defined. Should be left empty for people to fill themself.

Fixed that

> All sensors, except Absolute Humidity, are not associated with the Tempest device. They are in the system, but as separate sensors. This needs to be fixed.

Running a rebuild on mine again to see if I introduced something that broke it

natekspencer commented:
> All sensors, except Absolute Humidity, are not associated with the Tempest device. They are in the system, but as separate sensors. This needs to be fixed.

Try forcing a browser refresh. Mine did the same thing, but after refresh they all showed with the device.

briis commented:
I had left my computer for 10 min. when I came back they were all attached. Must be the internal workings of Home Assistant that introduces the delay.

Did not re-test the Area thing, but you wrote you fixed it. Even though that Backyard is a good guess on where the Device is placed, I am from Denmark, and I would call it something different. So better leave it blank - which is also the HA standard.

All looks good now.  So I think you can merge it. I will then release it tomorrow morning.

briis commented:
I have changed my mind with regards to the `wf_` prefix. This version introduces Breaking Changes already, so let us just take them all at once, and not do one more release with Breaking Changes.
So would you replacing the wf_ with the correct name?

natekspencer commented:
> I have changed my mind with regards to the `wf_` prefix. This version introduces Breaking Changes already, so let us just take them all at once, and not do one more release with Breaking Changes. So would mind you replacing the wf_ with the correct name?

Yup! I'll take care of it in a bit here!

briis commented:
Version 3.0.0 is now built and uploaded to docker. Thanks for this great work👏🏻
